src/mod_muc_room.erl

```diff
@@ -1953,6 +1976,14 @@ add_new_user(From, Nick, Packet, StateData) ->
 		  {error, Err}
 	  end;
       {_, _, _, Role} ->
+      RoomJID = StateData#state.jid,
+      To = jid:replace_resource(RoomJID, Nick),
+      case Role of
+      visitor ->
+        set_role(To, visitor, StateData);
```

Comment:
I kinda don't understand what this dose, you're not assigning the state data here

Reply:
This is the piece that persists visitor role status on room join. If you a join a room and your default role is visitor, assign role to visitor. (When affiliation is muted, default role = visitor)

set_role does end up assigning the state data `StateData#state{users = Users, nicks = Nicks}.`